### PR TITLE
[SPARK-11583][Scheduler][Core] Make MapStatus use less memory uage

### PR DESCRIPTION
**Pull request by yaooqinn**
Disscuss about this pr is at https://issues.apache.org/jira/browse/SPARK-11583

In the resolved issue https://issues.apache.org/jira/browse/SPARK-11271, as I said, using BitSet can save ≈20% memory usage compared to RoaringBitMap.
For a spark job contains quite a lot of tasks, 20% seems a drop in the ocean.
Essentially, BitSet uses long[]. For example a BitSet[200k] = long[3125].
So if we use a HashSet[Int] to store reduceId (when non-empty blocks are dense,use reduceId of empty blocks; when sparse, use non-empty ones).
For dense cases: if HashSet[Int](numNonEmptyBlocks).size < BitSet[totalBlockNum], I use MapStatusTrackingNoEmptyBlocks
For sparse cases: if HashSet[Int](numEmptyBlocks).size < BitSet[totalBlockNum], I use MapStatusTrackingEmptyBlocks
sparse case, 299/300 are empty
sc.makeRDD(1 to 30000, 3000).groupBy(x=>x).top(5)
dense case, no block is empty
sc.makeRDD(1 to 9000000, 3000).groupBy(x=>x).top(5)

### CODE REVIEW

**Comment by yaooqinn** [comment]
retest plz

**Comment by watermen** [comment]
retest this please

**Comment by rxin** [comment]
Can you use the OpenHashSet in Spark?

**Comment by andrewor14** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 1]
```diff
@@ -121,6 +123,65 @@ private[spark] class CompressedMapStatus(
...
+ * @param numNonEmptyBlocks the number of non-empty blocks
+ * @param markedBlocks a HashSet tracking which blocks are empty(dense) / non-empty(sparse)
+ * @param avgSize average size of the non-empty blocks
+ */
+private[spark] class MapStatusTrackingEmptyBlocks private (
+    private[this] var loc: BlockManagerId,
+    private[this] var numNonEmptyBlocks: Int,
+    private[this] var markedBlocks: mutable.HashSet[Int],
+    private[this] var avgSize: Long,
+    private[this] var isSparse: Boolean)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: why are these `var`s? You never set them to anything

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: you can just do
```
private[spark] class MapStatusTrackingEmptyBlocks private (
    loc: BlockManagerId,
    ...)
  extends MapStatus with ...
```
by default that's a private val, which is what you want

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 2]
```diff
@@ -121,6 +123,65 @@ private[spark] class CompressedMapStatus(
...
+  }
+}
+
+private[spark] object MapStatusTrackingEmptyBlocks {
+  def apply(
+    loc: BlockManagerId,
+    numNonEmptyBlocks: Int ,
+    markedBlocks: mutable.HashSet[Int],
+    avgSize: Long,
+    flag: Boolean): MapStatusTrackingEmptyBlocks = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: indent these by 2 more spaces

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 3]
```diff
@@ -141,7 +202,7 @@ private[spark] class HighlyCompressedMapStatus private (
     "Average size can only be zero for map stages that produced no output")
 
   protected def this() = this(null, -1, null, -1)  // For deserialization only
-
+
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: revert

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 4]
```diff
@@ -193,6 +258,16 @@ private[spark] object HighlyCompressedMapStatus {
     } else {
       0
     }
-    new HighlyCompressedMapStatus(loc, numNonEmptyBlocks, emptyBlocks, avgSize)
+    var isSparse = true
+    if(numNonEmptyBlocks * 32 < totalNumBlocks){
+      MapStatusTrackingEmptyBlocks(loc, numNonEmptyBlocks, nonEmptyBlocks, avgSize, isSparse)
+    }
+    else if ((totalNumBlocks - numNonEmptyBlocks) * 32 < totalNumBlocks){
+      isSparse = false
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: this use of a `var` is unnecessarily complicated. Just do
```
if (numNonEmptyBlocks * 32 < totalNumBlocks) {
  MapStatusTrackingEmptyBlocks(loc, numNonEmptyBlocks, nonEmptyBlocks, avgSize, isSparse = true)
} else if (...) {
  MapStatusTrackingEmptyBlocks(loc, numNonEmptyBlocks, nonEmptyBlocks, avgSize, isSparse = false)
} else {
  ...
}
```

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 5]
```diff
@@ -121,6 +123,65 @@ private[spark] class CompressedMapStatus(
...
+  override def writeExternal(out: ObjectOutput): Unit = Utils.tryOrIOException {
+    loc.writeExternal(out)
+    out.writeObject(markedBlocks)
+    out.writeLong(avgSize)
+  }
+
+  override def readExternal(in: ObjectInput): Unit = Utils.tryOrIOException {
+    loc = BlockManagerId(in)
+    markedBlocks = new mutable.HashSet[Int]
+    markedBlocks = in.readObject().asInstanceOf[mutable.HashSet[Int]]
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: did you mean to immediately override the assignment on the previous line?

**Comment by yaooqinn** [comment]
@rxin OpenHashSet replace HashSet

**Comment by yaooqinn** [comment]
@andrewor14 Thanks for your advices

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 6]
```diff
@@ -193,6 +254,12 @@ private[spark] object HighlyCompressedMapStatus {
     } else {
       0
     }
-    new HighlyCompressedMapStatus(loc, numNonEmptyBlocks, emptyBlocks, avgSize)
+    if(numNonEmptyBlocks * 32 < totalNumBlocks){
+      MapStatusTrackingEmptyBlocks(loc, numNonEmptyBlocks, nonEmptyBlocks, avgSize, isSparse = true)
+    } else if ((totalNumBlocks - numNonEmptyBlocks) * 32 < totalNumBlocks){
+      MapStatusTrackingEmptyBlocks(loc, numNonEmptyBlocks, emptyBlocksHashSet, avgSize, isSparse = false)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: `isSparse` is the wrong name, I think -- both cases are sparse, its a question of whether or not you are storing the empty blocks.

**Comment by squito** [comment]
Sean has raised some important higher level questions on the jira -- I'd like us to resolve the discussion there before moving forward on this.

**Comment by andrewor14** [comment]
@yaooqinn Can you close this now that you have the new patch?

**Comment by yaooqinn** [comment]
OK, close this pr and  see #9661

**Pull request closed without merging**